nexus/src/sagas.rs

```diff
@@ -514,6 +517,53 @@ async fn sic_create_instance_record(
     Ok(instance.runtime().clone().into())
 }
 
+async fn sic_delete_instance_record(
+    sagactx: ActionContext<SagaInstanceCreate>,
+) -> Result<(), anyhow::Error> {
+    println!("A");
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+    let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
+
+    // We currently only support deleting an instance if it is stopped or
+    // failed, so update the state accordingly to allow deletion.
+    let runtime_state =
+        sagactx.lookup::<InstanceRuntimeState>("initial_runtime")?;
+    let runtime_state = db::model::InstanceRuntimeState {
+        state: db::model::InstanceState::new(InstanceState::Failed),
+        // Must update the generation, or the database query will fail.
+        gen: db::model::Generation::from(runtime_state.gen.next()),
```

Comment:
This basically has the implicit assumption that "the only other thing which can move us away from the `Creating` state is either the saga completing or this undo action", right?

Reply:
That's correct. I tried to make that clear in the following block comment on the `assert!(updated, ...)` line. Though I like your phrasing: "This generation can only be moved forward by the saga's successful completion, or backward here."